lightning/src/util/persist.rs

```diff
@@ -249,10 +258,966 @@ where
+				if update_name.0 <= current_monitor.get_latest_update_id() {
+					self.kv_store.remove(
+						CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+						monitor_name.as_str(),
+						update_name.as_str(),
+						lazy,
+					)?;
+				}
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<
+		ChannelSigner: WriteableEcdsaChannelSigner,
+		K: Deref,
+		L: Deref,
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+	> Persist<ChannelSigner> for MonitorUpdatingPersister<K, L, ES, SP>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	/// Persists a new channel. This means writing the entire monitor to the
+	/// provided [`KVStore`].
+	fn persist_new_channel(
+		&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChannelSigner>,
+		_monitor_update_call_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		// Determine the proper key for this monitor
+		let monitor_name = MonitorName::from(funding_txo);
+		let maybe_old_monitor = self.read_monitor(&monitor_name);
+		match maybe_old_monitor {
+			Ok((_, ref old_monitor)) => {
+				// Check that this key isn't already storing a monitor with a higher update_id (collision)
+				if old_monitor.get_latest_update_id() > monitor.get_latest_update_id() {
+					log_error!(
+						self.logger,
+						"Tried to write a monitor at the same outpoint {} with a higher update_id!",
+						monitor_name.as_str()
+					);
+					return chain::ChannelMonitorUpdateStatus::UnrecoverableError;
+				}
+			}
+			// This means the channel monitor is new.
+			Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
+			_ => return chain::ChannelMonitorUpdateStatus::UnrecoverableError,
+		}
+		// Serialize and write the new monitor
+		let mut monitor_bytes = Vec::with_capacity(
+			MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() + monitor.serialized_length(),
+		);
+		monitor_bytes.extend_from_slice(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL);
+		monitor.write(&mut monitor_bytes).unwrap();
+		match self.kv_store.write(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			monitor_name.as_str(),
+			&monitor_bytes,
+		) {
+			Ok(_) => {
+				// Assess cleanup. Typically, we'll clean up only between the last two known full monitors.
+				if let Ok((_, old_monitor)) = maybe_old_monitor {
+					let start = old_monitor.get_latest_update_id();
+					let mut end = monitor.get_latest_update_id();
+					// We should bother cleaning up only if the update_id difference is greater
+					// than one; otherwise, it's implied that no updates were stored.
+					if end - start > 1 {
+						if end == CLOSED_CHANNEL_UPDATE_ID {
+							// We don't want to clean the rest of u64, so just do possible pending updates. Note that
+							// we never write updates at `CLOSED_CHANNEL_UPDATE_ID`.
+							end = start.saturating_add(self.maximum_pending_updates);
+						}
+						for update_id in start + 1..=end - 1 {
```

Comment:
Yeah, I recognized this, too. I experimented with some other variations locally, I didn't feel they were much less complex. One thing is that the old monitor could be closed, I believe (IIUC, we re-write monitors for closed channels sometimes) so that has to be a saturating add or something. 

Lemme take another crack at it.

Reply:
Okay, this is a bit easier to reason about, I hope. The saturating add/sub are required, I think, because we re-write monitors and both 0 and u64::MAX. I mean, I could probably do a checked operation instead, but it will start getting more complex.

```rust
// Assess cleanup. Typically, we'll clean up only between the last two known full monitors.
if let Ok((_, old_monitor)) = maybe_old_monitor {
	// Updates must be > the old monitor's update_id.
	let start = old_monitor.get_latest_update_id().saturating_add(1);
	let end = if monitor.get_latest_update_id() == CLOSED_CHANNEL_UPDATE_ID {
		// We don't want to clean the rest of u64, so just do possible pending updates. Note that
		// we never write updates at `CLOSED_CHANNEL_UPDATE_ID`.
		cmp::min(
			start.saturating_add(self.maximum_pending_updates),
			CLOSED_CHANNEL_UPDATE_ID - 1,
		)
	} else {
		monitor.get_latest_update_id().saturating_sub(1)
	};
	// We should bother cleaning up only if there's at least one update expected.
	if end >= start {
		for update_id in start..=end {
			// ... same as before
```